external-crates/move/crates/move-compiler/src/naming/translate.rs

```diff
@@ -1301,70 +1520,55 @@ fn exp_(context: &mut Context, e: E::Exp) -> N::Exp {
             }
         }
         EE::Mutate(el, er) => {
-            let nel = exp(context, *el);
-            let ner = exp(context, *er);
+            let nel = exp(context, el);
+            let ner = exp(context, er);
             NE::Mutate(nel, ner)
         }
 
-        EE::Abort(es) => NE::Abort(exp(context, *es)),
+        EE::Abort(es) => NE::Abort(exp(context, es)),
         EE::Return(name_opt, es) => {
-            let out_rhs = exp(context, *es);
+            let out_rhs = exp(context, es);
             if let Some(block_name) = name_opt {
-                if let Some(return_name) =
-                    context.resolve_nominal_label("return", NominalBlockType::Block, block_name)
-                {
-                    NE::Give(return_name, out_rhs)
-                } else {
-                    NE::UnresolvedError
-                }
+                context
+                    .resolve_nominal_label(NominalBlockUsage::Return, block_name)
+                    .map(|name| NE::Give(NominalBlockUsage::Return, name, out_rhs))
+                    .unwrap_or_else(|| NE::UnresolvedError)
+            } else if let Some(return_name) = context.current_return(eloc) {
+                NE::Give(NominalBlockUsage::Return, return_name, out_rhs)
             } else {
                 NE::Return(out_rhs)
             }
```

Comment:
Small nit, this may be easier to read as two split-out cases (like `Block` above):

```
EE::Return(Some(block_name), es) => {
    let out_rhs = exp(context, es);
    context
        .resolve_nominal_label(NominalBlockUsage::Return, block_name)
        .map(|name| NE::Give(NominalBlockUsage::Return, name, out_rhs))
        .unwrap_or_else(|| NE::UnresolvedError)
}
EE::Return(None, es) => {
    let out_rhs = exp(context, es);
    if let Some(return_name) = context.current_return(eloc) {
        NE::Give(NominalBlockUsage::Return, return_name, out_rhs)
    } else {
             NE::Return(out_rhs)
    }
}
```